lib/srv/sessiontracker_test.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSessionTracker(t *testing.T) {
+	ctx := context.Background()
+	clock := clockwork.NewFakeClockAt(time.Now())
+
+	mockService := &mockSessiontrackerService{
+		trackers: make(map[string]types.SessionTracker),
+	}
+
+	sessID := "sessionID"
+	trackerSpec := types.SessionTrackerSpecV1{
+		SessionID: sessID,
+		State:     types.SessionState_SessionStatePending,
+	}
+
+	// Create a new session tracker
+	tracker, err := NewSessionTracker(ctx, trackerSpec, mockService)
+	require.NoError(t, err)
+	require.NotNil(t, tracker)
+	require.Equal(t, tracker.tracker, mockService.trackers[sessID])
+
+	t.Run("UpdateExpirationLoop", func(t *testing.T) {
+		cancelCtx, cancel := context.WithCancel(ctx)
+		done := make(chan struct{})
+
+		// Test update expiration loop
+		go func() {
+			tracker.UpdateExpirationLoop(cancelCtx, clock)
+			close(done)
+		}()
+
+		// Wait for goroutine to wait on clock.After
+		clock.BlockUntil(1)
+		expectedExpiry := tracker.tracker.Expiry().Add(sessionTrackerExpirationUpdateInterval)
+		clock.Advance(sessionTrackerExpirationUpdateInterval)
+
+		// expiration should be updated by the next time we wait on clock.After
+		clock.BlockUntil(1)
+		require.Equal(t, expectedExpiry, tracker.tracker.Expiry())
+		require.Equal(t, tracker.tracker, mockService.trackers[sessID])
+
+		// cancelling the goroutine's ctx should halt the update loop
+		cancel()
+		require.Nil(t, <-done)
+	})
+
+	t.Run("WaitOnStateUpdate", func(t *testing.T) {
+		t.Parallel()
+
```

Comment:
Ditch? I don't think this would help much, plus it's a bit hard to reason about.

```suggestion
```

Reply:
Hmm I'm not sure there's another way to test `WaitOnStateChange`, but I can just remove the test if you think it's unnecessary.

Reply:
I think I misread this as an independent test scenario. If it does depend on something else, let's join the scenarios or fire up an explicit goroutine. `t.Parallel` is a hint, not an obligation by the framework.